### DISCUSSION TITLE
Could you share any papers/books that provide underlying explanations for this project?

### DISCUSSION THREAD

**Discussion by nicks64** [NVIDIA/TransformerEngine#807] [General]
Hi i am looking to contribute to this project. I am just getting started... and I am looking for papers/books that explain the logic/concept of this open source project. Could you share any and everything that i should look into before I jump in?

I know that attention is all you need is a good start.

Thank you in advance

**Comment by sbhavani** [comment]
I'd recommend starting with our GTC videos [What's New in Transformer Engine and FP8 Training | GTC 2024](https://www.nvidia.com/en-us/on-demand/session/gtc24-s62457/) and [FP8 Training with Transformer Engine | GTC 2023](https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51393). You can also run our FP8 examples on any Hopper or Ada GPU.

Once you're ready to contribute please refer to [CONTRIBUTING.rst](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst) guide.

**Reply by nicks64** [reply]
i personally think i need to look into mathematical concept to contribute... am i wrong?

or some books/research papers.

**Comment by timmoon10** [comment]
Transformer Engine is a library for accelerating Transformer models on NVIDIA GPUs. As you mention, ["Attention is all you need"](https://arxiv.org/abs/1706.03762) is a good introductory paper on this model architecture, and it may be helpful to study variations like [BERT](https://arxiv.org/pdf/1810.04805.pdf) or [GPT](https://www.mikecaptain.com/resources/pdf/GPT-1.pdf).

Our biggest feature is probably support for FP8 training, including tuned CUDA kernels, integrations into cuBLAS and cuDNN, and automatic handling of FP8 scaling factors in multiple DL frameworks. We use [the FP8 E4M3 and FP8 E5M2 formats](https://arxiv.org/pdf/2209.05433.pdf). It may be useful to look at [the Hopper white paper](https://resources.nvidia.com/en-us-tensor-core/gtc22-whitepaper-hopper) and [FP16 training](https://arxiv.org/pdf/1710.03740.pdf). I have also seen some interesting work on FP8 training by the [MS-AMP team](https://arxiv.org/pdf/2310.18313.pdf).

TE also integrates with the large-scale training techniques used in [Megatron-LM](https://github.com/NVIDIA/Megatron-LM) and [NeMo](https://github.com/NVIDIA/NeMo). Important topics include [data parallelism](https://arxiv.org/abs/1404.5997), [tensor parallelism](https://arxiv.org/abs/1909.08053), [ZeRO/FSDP](https://arxiv.org/abs/1910.02054), and [pipeline parallelism](https://arxiv.org/abs/1811.06965).

Finally, the best basic overviews are probably the [TE docs](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/index.html) and @ptrendx's [2023 GTC talk](https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51393/).

**Reply by nicks64** [reply]
i personally think i need to look into mathematical concepts to contribute... am i wrong?

or work on the project in tangent

or some books/research papers.
